Update check for Kirby and plugins

lukasbestle opened the pull request:
## Update check for Kirby and plugins

<img width="1107" alt="Screenshot" src="https://user-images.githubusercontent.com/1595007/192364766-3fce6de8-ea31-49db-909c-6dd94f5a74b1.png">

Our new update check in the enhanced system view makes it easy to stay informed about the version and security status of Kirby and your installed plugins. While feature updates are not always needed for finished sites, keeping an eye on security issues and important security messages is really important to keep your sites secure and healthy. We already publish all relevant information on our [security page](https://getkirby.com/security). The new update check brings this information and more right into the Panel so you can get a quick overview of the status of your particular site.

We designed the feature to work in your interest: It always prefers the minimum possible security update over feature updates to make security updates as easy as possible. It never performs an auto-update but just displays the relevant information. It never transmits any personal or site data to our server but processes the necessary data directly on your own server. And if you prefer to keep up to date with updates in another way, you can fine-tune the behavior or disable the feature all-together:

```php
<?php

// global configuration
return [
  'updates' => true | 'security' | false
];

// separate configuration for Kirby and plugins
return [
  'updates.kirby'   => true | 'security' | false,
  'updates.plugins' => true | false,
];

// or even configuration by plugin
return [
  'updates.plugins' => [
    '*'                      => true | false,
    'superwoman/*'           => true | false,
    'superwoman/superplugin' => true | false
  ]
];
```

---

## Related links

https://kirby.nolt.io/111

## Testing information

To test different scenarios, please modify the version number in `kirby/composer.json`. Changes to the version number will be picked up immediately. After a reload of the system view you can verify if the output matches what you expect.

Automated tests are still missing for parts of the internal update check logic. I will add complete unit and integration tests for all the logic until the 3.8 release. My goal is to write detailed integration tests with lots of different raw inputs and their expected outputs, including edge cases.

## Open questions

- [x] I'm not 100 % happy with the `TableUpdateStatusCell.vue` component yet. Its logic works well, but it could use better styling. I'm mostly not sure about the link styling: The underline is consistent with `UrlFieldPreview`, but the visual semantics differ from the Kirby update status in the `k-stats`. In both cases the link leads to the release notes of the update (if an update is available) or of the current version (if no update is available), but the stat (Kirby) & table cell (plugin) always display the current version. So I'm wondering if we should make the whole table cell clickable. Then it's more clear that the link is about the version and its update, not just the current version.
- [ ] The plan is to use `https://assets.getkirby.com` (KeyCDN) to serve the update files. We need to verify that we can indeed use KeyCDN for this (bandwidth, cost, caching behavior...).

## Ready?
<!--
If you can help to check off the following tasks, that'd be great.
If not, don't worry - we will take care of it.

More details: https://contribute.getkirby.com
-->

- [ ] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] Tests and checks all pass


### For review team
<!-- 
We will take care of the following before merging the PR.
-->

- [x] Add changes to release notes draft in Notion
- [x] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)
  - `update` option
  - Privacy policy
  - Privacy guide

lukasbestle committed:
Add `composer/semver` dependency

lukasbestle committed:
System cache `updates`

lukasbestle committed:
System view: Support for different icons per issue

lukasbestle committed:
New class `Kirby\Cms\System\UpdateStatus`

lukasbestle committed:
Display update status on system view

lukasbestle committed:
`updateStatus()` methods for system and plugins

distantnative commented on config/areas/system/views.php:
```diff
@@ -21,7 +22,13 @@
+					'link'  => (
+						$updateStatus ?
+						$updateStatus->url() :
+						'https://github.com/getkirby/kirby/releases/tag/' . $kirby->version()
```
```php
$updateStatus?->url() ?? 'https://github.com/getkirby/kirby/releases/tag/' . $kirby->version()
```

lukasbestle replied:
This won't work because `$updateStatus->url()` can also return `null` (for "no URL"). We only need the fallback if the whole `$updateStatus` object is `null`.

distantnative commented on config/areas/system/views.php:
```diff
@@ -35,18 +42,24 @@
+					$plugin->version() ?
+					$plugin->updateStatus()?->toArray() :
+					null
+				) ?? $plugin->version() ?? '–';
```
Maybe it's too late today for me to understand this: But aren't we already checking in the parentheses for `$plugin->version()`? But if that exists we go for `$plugin->updateStatus()`. But if `$plugin->version()` return null, we jump to `null` in the parentheses just to fallback to `$plugin->version()` outside again?

lukasbestle replied:
You are right, that makes no sense. I'll double-check what I intended to do.

lukasbestle replied:
@bastianallgeier I know your rewrite is not done yet, just before I forget about it: The general fallback to `$plugin->version()` is needed when the update check is disabled for one plugin or all plugins. In this case the `$plugin->updateStatus()` is `null`, so the Vue component can only get the version number as string. In your current version, support for the fallback is completely removed.

lukasbestle replied:
@distantnative Reading my original code again, it made sense, but is confusing. If the version is there but `$plugin->updateStatus()` returns `null` (= update check disabled), it used `$plugin->version()` as fallback. This is the part that's missing now with the new code.

distantnative replied:
@lukasbestle you know I'm not an `else` fan :P But in this case I think an if-else statement is in order. trying to make this a one liner really hinders understanding it.

lukasbestle replied:
On the upside, the current code has the advantage that it displays the latest version even if there is no current version.

So I think the ideal way would be this:

```php
$version = $plugin->updateStatus()?->toArray() ?? $plugin->version() ?? '?';
```

And then of course the frontend component needs to deal with the string and only display a bubble without dropdown or something along those lines.

distantnative commented on i18n/translations/en.json:
```diff
@@ -487,11 +487,23 @@
 
   "stats.empty": "No reports",
   "system.issues.content": "The content folder seems to be exposed",
+  "system.issues.eol.kirby": "Your installed Kirby version is end of life and will not receive further security updates",
```
```suggestion
  "system.issues.eol.kirby": "Your installed Kirby version has reached end-of-life and will not receive further security updates",
```

distantnative commented on i18n/translations/en.json:
```diff
@@ -487,11 +487,23 @@
   "stats.empty": "No reports",
   "system.issues.content": "The content folder seems to be exposed",
+  "system.issues.eol.kirby": "Your installed Kirby version is end of life and will not receive further security updates",
+  "system.issues.eol.plugin": "Your installed version of the { plugin } plugin is end of life and will not receive further security updates",
```
```suggestion
  "system.issues.eol.plugin": "Your installed version of the { plugin } plugin is has reached end-of-life and will not receive further security updates",
```

distantnative commented on panel/src/components/Layout/TableUpdateStatusCell.vue:
```diff
@@ -0,0 +1,41 @@
+		},
+		version() {
+			return this.hasInfo ? this.value.version : this.value;
+		}
```
I think we can spare `hasInfo`. On the `v-if` we can directly check for `value.label`. And in the computed props we can directly return `this.value.url ?? null` and `this.value.version ?? this.value`.

lukasbestle replied:
Certainly an option. I'm just wondering if it is too clever. The `hasInfo` value makes it easy to see that the cell can have two modes (strings and a whole object).

distantnative commented on panel/src/config/components.js:
```diff
@@ -108,6 +108,7 @@ import Panel from "@/components/Layout/Panel.vue";
 import Stats from "@/components/Layout/Stats.vue";
 import Table from "@/components/Layout/Table.vue";
 import TableCell from "@/components/Layout/TableCell.vue";
+import TableUpdateStatusCell from "@/components/Layout/TableUpdateStatusCell.vue";
```
This should probably rather go into `Forms/Previews` like other table cell previews.

lukasbestle replied:
It's not a field preview though, is it? `TableCell.vue` supports either table cell components or field preview components. I used the former here.

distantnative commented on src/Cms/Plugin.php:
```diff
@@ -212,6 +214,53 @@ public function toArray(): array
+	 *
+	 * @param array|null $data Custom override for the getkirby.com update data
+	 */
+	public function updateStatus(array|null $data = null): UpdateStatus|null
```
Wondering if `updates()` would be a more Kirby-like name and still fitting.

lukasbestle replied:
Thought about it as well. I find `updates()` a bit weird because there can only be a single update. But a singular `update()` sounds like a verb. So I decided on `updateStatus()` even though it's also not perfect.

distantnative replied:
I was thinking again about it, but is it really a single update when I'm on version 3.7.1 and the system gives me info on updating to 3.7.5 - or is that four updates?

lukasbestle replied:
The class only ever suggests a single update. It is the user's responsibility to decide whether intermediate steps are useful and/or necessary.

distantnative replied:
It suggests a version number, right? My point was that this would/could suggest multiple updates - which would make the name not that strange at all.

lukasbestle replied:
Suggesting multiple updates would require a rewrite because there can only be one return value for each of the methods. Also I feel like multiple updates would be too overwhelming (e.g. if there is a security update, the user shouldn't need to care that there is also a feature update and an upgrade available – until that security update is done).

distantnative replied:
Again, I'm not talking about changing anything but the name. Returning an info about 3.7.5 when you're on 3.7.1 is info about updates, not just one. And since I think the current name is a bit cumbersome, my only argument here is to use `Updates` and that it does make sense. Not to change anything how it works or what is returned.

distantnative replied:
But if you like the current name that much, fine, we can stick with it.

lukasbestle replied:
I don't like either name 100 %. :)

bastianallgeier replied:
Let's stay with updateStatus for now.

distantnative commented on src/Cms/Plugin.php:
```diff
@@ -212,6 +214,53 @@ public function toArray(): array
+				$option,
+				fn ($value, $key) => fnmatch($key, $this->name()) === true,
+				ARRAY_FILTER_USE_BOTH
+			);
```
We could use `A::filter` here

distantnative commented on src/Cms/System/UpdateStatus.php:
```diff
@@ -0,0 +1,680 @@
+			return $this->vulnerabilities;
+		}
+
+		if (!$this->data || !$this->currentVersion) {
```
```suggestion
		if ($this->data === null || $this->currentVersion === null) {
```

Let's be explicit in the if statements, the `!` is always hard to understand what it really is supposed to test for.

lukasbestle replied:
To be honest my goal with these checks was to be as lax as possible explicitly. If the value ends up to be falsy but not null for some reason, we still need to return early. Checking strictly for null would be a downgrade from that safety. We do have prop and return types everywhere, but this provides an extra level of safety.

I can still change it, but I think it's another one of these general decisions we need to make as a team.

distantnative replied:
Agreed, should be a common decision we all stick too - I thought we had made this one with v3, but happy to revisit and renew or change it. 

My argument for making it explicit would be readability and by that making our code easier to understand for ourselves. For you, this might be a given as you wrote that code. But for us others on the team things like this make it harder to understand as it opens up a lot of options. Given that `data` can only be an array or null, and `currentVersion` only a string or null, I think the openness doesn't add a lot more functionality, just decreases readability. For me to understand this, I had to jump to the very top of the file to read the property types, just to jump back down again. Would be much easier if I could understand that code in place.

lukasbestle replied:
IMO the possible types don't really matter here. It's just "if this value is not set, then we cannot continue and need to return early".

distantnative reviewed (commented):
Some first comments - I have given up to review more tonight, realising how complex the new class is. Will try to have a closer look with a fresh mind.

On the config options, I think if the set is 
```
'display' | 'display-security' | false,
```
we might want to simplify it to align it with more of our other options and do
```
true | 'security' | false
```

I do agree on the styling of that cell, but that's usually best left with @bastianallgeier.

lukasbestle committed:
Reword EOL warnings

Co-authored-by: Nico Hoffmann  ෴. <me@nhoffmann.com>

lukasbestle committed:
Merge UpdateStatus from improvements branch

lukasbestle committed:
Use `A::filter()`

lukasbestle committed:
More explicit empty value checking

lukasbestle committed:
Rename option from `update` to `updates`

lukasbestle committed:
Option values `true | 'security' | false`

lukasbestle committed:
Plugins: No security-only mode for now

lukasbestle committed:
Display latest version even w/o current version

lukasbestle committed:
Merge branch 'develop' into feature/update-check

# Conflicts:
#	i18n/translations/en.json

lukasbestle commented on src/Cms/System/UpdateStatus.php:
```diff
@@ -0,0 +1,708 @@
+			'theme'         => $this->theme(),
+			'url'           => $this->url(),
+			'version'       => $this->currentVersion ?? '?',
+			'targetVersion' => $this->targetVersion() ?? $this->currentVersion ?? '?'
```
@bastianallgeier I guess you added this fallback to the current version to display the latest version even if the version is already up to date, right?

There is also the case of an unreleased version (current version is higher than the latest version), in this case the fallback will break. If we want to display the latest version, we can get it with `$this->data['latest'] ?? null` more reliably.

Suggestion: What if the dropdown displays three versions like this (now that we have the room):

<img src="https://user-images.githubusercontent.com/1595007/191598661-27631f6b-f838-4f12-89ee-cb92e6cf8696.png" alt="Screenshot" width="320">

The data can come from here:

```suggestion
			'currentVersion' => $this->currentVersion ?? '?',
			'latestVersion'  => $this->data['latest'] ?? '?',
			'targetVersion'  => $this->targetVersion()
```

`targetVersion` explicitly has *no* fallback. If it's null, the dropdown would hide that "Suggested version" line entirely and only display current and latest version.

lukasbestle replied:
Thinking about it, the suggested version is never needed as it's already contained in the status label. So I think current version and latest version are enough:

```suggestion
			'currentVersion' => $this->currentVersion ?? '?',
			'latestVersion'  => $this->data['latest'] ?? '?',
```

lukasbestle committed:
Don't suggest a target version on error

lukasbestle committed:
Fix typo

lukasbestle committed:
Merge branch 'develop' into feature/update-check

# Conflicts:
#	composer.lock

lukasbestle committed:
Merge in draft for improved update status cell

Includes reliability improvements regarding the latest version and disabled update check

lukasbestle committed:
Simple text fallback if update check is disabled

lukasbestle committed:
Get rid of hasInfo prop

lukasbestle committed:
Disable version information button for missing URL

bastianallgeier committed:
Slightly nicer green

bastianallgeier committed:
Auto-height for stats cells

bastianallgeier committed:
Fixed color for url previews without links

bastianallgeier committed:
Show message below unregistered license in system view

bastianallgeier committed:
Less aggressive colors in system stats

bastianallgeier committed:
Improve design of the system stats

lukasbestle committed:
Fix coding style

lukasbestle committed:
Update status dropdown: Translatable labels

lukasbestle committed:
UpdateStatus: Use `pluginName` prop naming

lukasbestle committed:
Fix unit tests

lukasbestle committed:
Unit tests for `System::updateStatus()`

lukasbestle committed:
Add `@since` tags

lukasbestle committed:
System view: Print update check errors to console

lukasbestle committed:
Initialize exceptions after running logic

Ensures that the exceptions are complete (including processing exceptions, not just loading exceptions)

lukasbestle committed:
Cleanup

lukasbestle committed:
Minor fixes

lukasbestle committed:
More unit tests

afbora reviewed (approved):
Looks amazing ❤️
